lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,773 @@
+
+	fn do_local_state_transition(&mut self) -> Result<InteractiveTxMessageSend, ()> {
+		if let Some((input, prev_tx)) = self.inputs_to_contribute.pop() {
+			let msg = msgs::TxAddInput {
+				channel_id: ChannelId(self.channel_id),
+				serial_id: self.generate_local_serial_id(),
+				prevtx: TransactionU16LenLimited(prev_tx),
+				prevtx_out: input.previous_output.vout,
+				sequence: input.sequence.to_consensus_u32(),
+			};
+			let _ = do_state_transition!(self, local_tx_add_input, &msg)?;
+			Ok(InteractiveTxMessageSend::TxAddInput(msg))
+		} else if let Some(output) = self.outputs_to_contribute.pop() {
+			let msg = msgs::TxAddOutput {
+				channel_id: ChannelId(self.channel_id),
+				serial_id: self.generate_local_serial_id(),
+				sats: output.value,
+				script: output.script_pubkey,
+			};
+			let _ = do_state_transition!(self, local_tx_add_output, &msg)?;
+			Ok(InteractiveTxMessageSend::TxAddOutput(msg))
+		} else {
+			let msg = msgs::TxComplete { channel_id: ChannelId(self.channel_id) };
+			let _ = do_state_transition!(self, local_tx_complete, &msg)?;
+			Ok(InteractiveTxMessageSend::TxComplete(msg))
+		}
+	}
+
+	pub fn handle_tx_add_input(&mut self, msg: &msgs::TxAddInput) -> Result<InteractiveTxMessageSend, ()> {
+		let _ = do_state_transition!(self, remote_tx_add_input, msg)?;
+		self.do_local_state_transition()
+	}
+
+	pub fn handle_tx_remove_input(&mut self, msg: &msgs::TxRemoveInput) -> Result<InteractiveTxMessageSend, ()> {
+		let _ = do_state_transition!(self, remote_tx_remove_input, msg)?;
+		self.do_local_state_transition()
+	}
+
+	pub fn handle_tx_add_output(&mut self, msg: &msgs::TxAddOutput) -> Result<InteractiveTxMessageSend, ()> {
+		let _ = do_state_transition!(self, remote_tx_add_output, msg)?;
+		self.do_local_state_transition()
+	}
+
+	pub fn handle_tx_remove_output(&mut self, msg: &msgs::TxRemoveOutput) -> Result<InteractiveTxMessageSend, ()> {
+		let _ = do_state_transition!(self, remote_tx_remove_output, msg)?;
+		self.do_local_state_transition()
+	}
+
+	pub fn handle_tx_complete(&mut self, msg: &msgs::TxComplete) -> Result<(Option<InteractiveTxMessageSend>, Option<Transaction>), ()> {
+		let _ = do_state_transition!(self, remote_tx_complete, msg)?;
+		match &self.state_machine {
+			StateMachine::RemoteTxComplete(_) => {
+				let msg_send = self.do_local_state_transition()?;
+				let negotiated_tx = match &self.state_machine {
+					StateMachine::NegotiationComplete(s) => Some(s.0.clone()),
+					StateMachine::LocalChange(_) => None, // We either had an input or output to contribute.
+					_ => {
+						debug_assert!(false, "We cannot transition to any other states after receiving `tx_complete` and responding");
+						return Err(());
+					}
+				};
+				Ok((Some(msg_send), negotiated_tx))
+			}
+			StateMachine::NegotiationComplete(s) => Ok((None, Some(s.0.clone()))),
+			_ => {
+				debug_assert!(false, "We cannot transition to any other states after receiving `tx_complete`");
+				Err(())
+			}
+		}
+	}
```

Comment:
I think `handle_abort` is still missing

Reply:
the holder of the interactive transaction constructor (`Channel`) would be responsible for this. specifically, it would hold an `Option<InteractiveTxConstructor>` and would just have to set it to `None`.